opentelemetry tracer: add support for environment resource detector

joaopgrassi opened the pull request:
**Commit Message**: Allow specifying resource detectors for the OpenTelemetry tracer via a new configuration `resource_detectors`. The resource detector reads from the env variable `OTEL_RESOURCE_ATTRIBUTES` which is defined by the [OTel specification](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.24.0/specification/resource/sdk.md#detecting-resource-information-from-the-environment). The detector returns a resource object populated with the detected attributes, which is sent as part of the OTLP request. 

**Additional Description**: This PR adds the "foundation" for building other resource detectors in Envoy. It is based on the [OTel collector implementation](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/v0.84.0/processor/resourcedetectionprocessor). Users can configure multiple resource detectors, and they work together to "merge" all the detected attributes into a single resource object, which is then part of the OTLP message exported.

**Risk Level**: Low

**Testing**: Multiple unit tests, that cover all new code/scenarios. I also did manual testing, running Envoy locally with the OTel tracer + env resource detector enabled. Resource attributes detected from my environment is successfully exported as seen in the Jaeger screenshot. 
<img width="1131" alt="resource-detectors-env-jaeger" src="https://github.com/envoyproxy/envoy/assets/5938087/9234d4e7-ce2c-45b2-81db-4525011eb586">


Docs Changes: Not sure if I should add/where. Happy to do it. 
Release Notes: N/A
Platform Specific Features: N/A
[Optional Runtime guard:] N/A
[Optional Fixes #28929] 

Here is how the new config is used:

```yaml
tracing:
  provider:
    name: envoy.tracers.opentelemetry
    typed_config:
      "@type": type.googleapis.com/envoy.config.trace.v3.OpenTelemetryConfig
       grpc_service:
         envoy_grpc:
           cluster_name: opentelemetry_collector
         timeout: 0.250s
       service_name: envoy-gRPC-exporter
       resource_detectors: # --> NEW CONFIG
         - name: envoy.tracers.opentelemetry.resource_detectors.environment
```

joaopgrassi commented on api/envoy/config/trace/v3/opentelemetry.proto:
```diff
@@ -25,4 +38,8 @@ message OpenTelemetryConfig {
+
+  // An ordered list of resource detectors
+  // [#extension-category: envoy.tracers.opentelemetry.resource_detectors]
+  repeated ResourceDetector resource_detectors = 3;
```
I saw that we have [TypedExtensionConfig](https://github.com/envoyproxy/envoy/blob/v1.27.0/api/envoy/config/core/v3/extension.proto) for this purpose and I wanted to use it. The problem here is that the field `typed_config` is required. for resource detectors, it's not always that it will need its own extra config, which is the case of the environment detector in this PR. 

To circumvent this, I created this new type `ResourceDetector` which is basically the same, but does not require the `typed_config`. @htuch @KBaichoo tagging you since you commented on the issue.

joaopgrassi replied:
I'm working on a next detector for Dynatrace, and that one will have its own configuration.

joaopgrassi replied:
Ah neat, that make thing much simpler I think. I will adapt this way, thank you!

joaopgrassi replied:
Done in 277d9ce35c173dc2340e5ff452593b8cefd0250e. Simplified the code a bit :)

joaopgrassi commented on source/extensions/tracers/opentelemetry/resource_detectors/environment/environment_resource_detector.h:
```diff
@@ -0,0 +1,34 @@
+
+private:
+  Server::Configuration::TracerFactoryContext&
+      context_; // TODO, is keeping a reference ok (ownership)?
```
I was not sure if it's OK to keep this reference in the detector. The detector needs it to access the `api()` method, which is then used on the `DataSource` to read the env variable. It will also be needed in other detectors, that read from a file for example.

joaopgrassi commented on source/extensions/tracers/opentelemetry/resource_detectors/environment/BUILD:
```diff
@@ -0,0 +1,31 @@
+load(
```
The idea I had is to allow each vendor/interested party to add their custom detectors here. 

```
resource_detectors
  - environment
  - dynatrace
  - aws
  - azure
  - gcp
  -...
```

joaopgrassi replied:
Ah yes, each would be under their own directory. Sorry that wasn't clear from my comment.

adisuissa reviewed (commented):
Thank you for your contribution.
Here's a first-pass review.
FWIW, it is better to create s series of smaller PRs to increase review speed.

joaopgrassi commented on source/extensions/tracers/opentelemetry/config.cc:
```diff
@@ -18,7 +19,9 @@ OpenTelemetryTracerFactory::OpenTelemetryTracerFactory()
     Server::Configuration::TracerFactoryContext& context) {
-  return std::make_shared<Driver>(proto_config, context);
+
+  ResourceProviderImpl resource_provider;
```
Basically I changed the Driver constructor to ask for a ResourceProvider to be able to mock/test it.

`opentelemetry_tracer_impl_test.cc` is more  like an integration test where we can see the final OTLP message and I wanted to check the attributes from the detectors are ending up there as they should.

> Consider using 2 Driver c'tors - one that receives a resource-provider and the other that calls it with a default resource provider.

Not sure I got the idea, but did you maybe mean to have the ctor that receives the detector and use that for the tests? And the "production" code would call the Driver ctor without it, meaning the default is always used?

joaopgrassi replied:
I created now a new constructor and changed it a bit: 62365cfbda23266f57308e2bdf2ef48cd8679baa. Let me know if that's what you had in mind.

joaopgrassi commented on source/extensions/tracers/opentelemetry/resource_detectors/environment/environment_resource_detector.cc:
```diff
@@ -0,0 +1,56 @@
+ * @return Resource A resource with the attributes from the OTEL_RESOURCE_ATTRIBUTES environment
+ * variable.
+ */
+Resource EnvironmentResourceDetector::detect() {
```
Yes that is correct. It runs right before the Driver is instantiated.

joaopgrassi commented on source/extensions/tracers/opentelemetry/resource_detectors/environment/environment_resource_detector.cc:
```diff
@@ -0,0 +1,56 @@
+  resource.schemaUrl = "";
+
+  TRY_NEEDS_AUDIT {
+    auto attributes_str = Config::DataSource::read(ds, true, context_.serverFactoryContext().api());
```
Looking at the `DataSource::read` it says

> * @throw EnvoyException if no DataSource case is specified and !allow_empty.

This means then that no exception will be thrown ever - it specifies the case (by calling `set_environment_variable`) + `allow_empty` is `true`. So I _think_ it's safe to remove the try/catch.

joaopgrassi replied:
I removed the try-catch in [40c4373](https://github.com/envoyproxy/envoy/pull/29547/commits/40c43739775038c1a4a281ca204050a480f23f4b)

joaopgrassi replied:
So, I was wrong. If the env variable is not set, it still throws an exception, which made my test fail. I went back with your initial suggestion and reduced the try-catch to only this line. [e3b92cb](https://github.com/envoyproxy/envoy/pull/29547/commits/e3b92cb6fad78cb4b9d0c106f8ef80aa8973dfbd)

joaopgrassi commented on source/extensions/tracers/opentelemetry/resource_detectors/environment/environment_resource_detector.cc:
```diff
@@ -0,0 +1,56 @@
+    }
+  }
+  END_TRY catch (const EnvoyException& e) {
+    ENVOY_LOG(error, "Failed to read resource attributes: {}.", e.what());
```
Makes sense. I commented [above](https://github.com/envoyproxy/envoy/pull/29547#discussion_r1324599329), I think we don't need the try-catch at all.

joaopgrassi commented on source/extensions/tracers/opentelemetry/resource_detectors/environment/environment_resource_detector.cc:
```diff
@@ -0,0 +1,56 @@
+    ENVOY_LOG(error, "Failed to read resource attributes: {}.", e.what());
+  }
+
+  return resource;
```
I wanted to not fail and stop Envoy from starting because today, we only load the "resource" with `service.name`. If now I change it to stop initialization I feel this would be a breaking change in behavior somehow.  

This is because there might be cases where a failure in enrichment is OK and not a "stop the world" situation. I will give an example from Dynatrace. Depending where/how the application is deployed, our agent "put" certain files with a bunch of attributes it gathers. The Dynatrace detector then tries to read from all of these files, since it does not know where the app is deployed (for ex in K8s). So it's expected that one of the files will not be present, if the app is say, not running in k8s.

Another example for this detector is maybe the env variable is not set (or has a typo). I don't think this should cause Envoy to exit. We can at maximum log a warning that this happened.

Additionally, in the OTel collector, all of the detectors return an empty resource if they encounter an error:

- [Azure](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/resourcedetectionprocessor/internal/azure/azure.go#L49)
- [Environment](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/resourcedetectionprocessor/internal/env/env.go#L58)
- [System](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/resourcedetectionprocessor/internal/system/system.go#L64)
- [Docker](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/resourcedetectionprocessor/internal/docker/docker.go#L52)

joaopgrassi replied:
>OOC, and apologies for my lack of OTel understanding

First, not a problem at all! You also help me on my lack of Envoy understanding :D 

> but if there's only a warning somewhere, due to misconfiguration, then:
> 
>  1. does it require an operator to look at the logs of of all deployed K8s to see which succeeded and which failed?
> 

That would be one option, but the spans are exported to a back-end and it will be obvious that the service is missing resource attributes, as they are an "essential part" in defining the service. So, I imagine if, for example, they don't set the environment variable where the attributes are, and when they are looking at spans/querying they won't find it for N services and that will be the indication that something is not right.

>  2. If an error occurs, is a default "Resource" is used or is the previously configured "Resource" is used?

Not sure I follow, but the Tracer/Driver is configured when Envoy starts up and it's not changed at runtime IIUC right? 
If that's true, then when a detector fails to "detect", an empty resource is returned, which is then can be merged with other resources detectors the user has configured. 

For ex, I have resource detector A configured and working. Then I configured another resource detector B. When envoy starts up, detector A works but B fails. Envoy will then start with resource A only.

For this specific resource detector that looks at the environment variable, the OTel C++ SDK does also return empty, for ex if the env variable is not set https://github.com/open-telemetry/opentelemetry-cpp/blob/ad626cec4598165e0d297b0bdb3880968c5dfe88/sdk/src/resource/resource_detector.cc.

adisuissa replied:
Thanks for explaining this. I've got a few followup questions here:

> > but if there's only a warning somewhere, due to misconfiguration, then:
> > 
> > 1. does it require an operator to look at the logs of of all deployed K8s to see which succeeded and which failed?
> 
> That would be one option, but the spans are exported to a back-end and it will be obvious that the service is missing resource attributes, as they are an "essential part" in defining the service. So, I imagine if, for example, they don't set the environment variable where the attributes are, and when they are looking at spans/querying they won't find it for N services and that will be the indication that something is not right.
> 
I'm assuming there will be cases where Envoy is operated by one entity, and the target servers that receive the spans will be operated by another entity. In that case, one may send an invalid configuration to Envoy, not knowing that it was invalid, and not receive the expected behavior in the target servers.
If the config is accepted, when trying to answer the question "where is the problem?", it could either be the component that sends Envoy the config, or in Envoy, or on the way back to the target servers. One will need to look at the logs of all these, in order to understand where the problem is.
If, however, the config is rejected, this should raise an issue, because the config-server entity will receive a NACK.

> > 2. If an error occurs, is a default "Resource" is used or is the previously configured "Resource" is used?
> 
> Not sure I follow, but the Tracer/Driver is configured when Envoy starts up and it's not changed at runtime IIUC right? If that's true, then when a detector fails to "detect", an empty resource is returned, which then can be merged with other resources detectors the user has configured.
> 
I think that a tracer/driver can be updated dynamically, not only during startup.
Envoy allows for dynamic config update using the [xds protocol](https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#xds-rest-and-grpc-protocol).

> For ex, I have resource detector A configured and working. Then I configured another resource detector B. When envoy starts up, detector A works but B fails. Envoy will then start with resource A only.
> 
> For this specific resource detector that looks at the environment variable, the OTel C++ SDK does also return empty, for ex if the env variable is not set https://github.com/open-telemetry/opentelemetry-cpp/blob/ad626cec4598165e0d297b0bdb3880968c5dfe88/sdk/src/resource/resource_detector.cc.

joaopgrassi replied:
@adisuissa  I understand your points. I guess we just need to decide in which approach we want to adopt here. 

As I explained, some resource detectors may "fail" (or be unable) to detect attributes and that is not to be considered an error (the case with Dynatrace I explained is one example). For this case, the spec is clear: (emphasis on this part)

> Errors should be handled as specified in the [Error Handling principles](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.24.0/specification/error-handling.md#basic-error-handling-principles). **Note the failure to detect any resource information MUST NOT be considered an error**, whereas an error that occurs during an attempt to detect resource information SHOULD be considered an error.

But the [Error Handling principles](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.24.0/specification/error-handling.md#basic-error-handling-principles) says:

>The API or SDK MAY fail fast and cause the application to fail on initialization, e.g. because of a bad user config or environment, but MUST NOT cause the application to fail later at run time, e.g. due to dynamic config settings received from the Collector.

Which aligns with what you are saying about Envoy - bad configuration at bootstrap of Envoy fails but at dynamic changes won't stop it. 

For the environment resource detector, if you think it aligns best with Envoy's practices, I can make it fail, for example if the resource detector was configured, but the Env variable is not present OR if the format of the data in the env variable is wrong. But I think this decision to "fail or not and stop Envoy" may be changing depending of the "error" or situation a resource detector has.

joaopgrassi replied:
Ok! I will change it to reject the config if:

- Env variable is not present
- Env variable is present but empty
- ( will try if possible) Env variable is present but with values with "wrong" format

joaopgrassi replied:
Done in [d20885e](https://github.com/envoyproxy/envoy/pull/29547/commits/d20885ea2dc178424a9b7ed30c61240beb78591d)

Now if the cases above happen, Envoy will not start.

joaopgrassi commented on source/extensions/tracers/opentelemetry/resource_detectors/resource_provider.cc:
```diff
@@ -0,0 +1,109 @@
+
+  Resource resource = createInitialResource(opentelemetry_config.service_name());
+
+  auto detectors_configs = opentelemetry_config.resource_detectors();
```
Done in [40c4373](https://github.com/envoyproxy/envoy/pull/29547/commits/40c43739775038c1a4a281ca204050a480f23f4b)

joaopgrassi commented on source/extensions/tracers/opentelemetry/resource_detectors/resource_provider.cc:
```diff
@@ -0,0 +1,109 @@
+}
+} // namespace
+
+const Resource ResourceProviderImpl::getResource(
```
Done in [40c4373](https://github.com/envoyproxy/envoy/pull/29547/commits/40c43739775038c1a4a281ca204050a480f23f4b)

joaopgrassi commented on source/extensions/tracers/opentelemetry/resource_detectors/resource_provider.cc:
```diff
@@ -0,0 +1,109 @@
+ * @param updating_schema_url The updating resource's schema URL.
+ * @return std::string The calculated schema URL.
+ */
+std::string mergeSchemaUrl(const std::string& old_schema_url,
```
Done in [40c4373](https://github.com/envoyproxy/envoy/pull/29547/commits/40c43739775038c1a4a281ca204050a480f23f4b). I used `resolveSchemaUrl`

joaopgrassi commented on source/extensions/tracers/opentelemetry/resource_detectors/resource_detector.h:
```diff
@@ -0,0 +1,69 @@
+   * populated with them and a possible SchemaUrl.
+   * @return Resource
+   */
+  virtual Resource detect() = 0;
```
Done in [40c4373](https://github.com/envoyproxy/envoy/pull/29547/commits/40c43739775038c1a4a281ca204050a480f23f4b)

joaopgrassi commented on source/extensions/tracers/opentelemetry/resource_detectors/resource_detector.h:
```diff
@@ -0,0 +1,69 @@
+
+using ResourceAttributes = std::map<std::string, std::string>;
+
+class Resource {
```
Done in [40c4373](https://github.com/envoyproxy/envoy/pull/29547/commits/40c43739775038c1a4a281ca204050a480f23f4b)

joaopgrassi commented on source/extensions/tracers/opentelemetry/resource_detectors/resource_detector.h:
```diff
@@ -0,0 +1,69 @@
+class Resource {
+public:
+  std::string schemaUrl{""};
+  ResourceAttributes attributes{};
```
Done in [40c4373](https://github.com/envoyproxy/envoy/pull/29547/commits/40c43739775038c1a4a281ca204050a480f23f4b)

joaopgrassi commented on source/extensions/tracers/opentelemetry/opentelemetry_tracer_impl.cc:
```diff
@@ -21,13 +23,17 @@ namespace Tracers {
   // Create the tracer in Thread Local Storage.
-  tls_slot_ptr_->set([opentelemetry_config, &factory_context, this](Event::Dispatcher& dispatcher) {
+  tls_slot_ptr_->set([opentelemetry_config, &factory_context, this,
+                      resource](Event::Dispatcher& dispatcher) {
```
Thanks for pointing this out, C++ is new to me still. I made it a shared pointer in [0febb7c](https://github.com/envoyproxy/envoy/pull/29547/commits/0febb7c2f831e1eb6ca81405f1aa8a089080c8b7). Let me know if you think this is reasonable!

joaopgrassi commented on source/extensions/tracers/opentelemetry/resource_detectors/resource_provider.cc:
```diff
@@ -0,0 +1,109 @@
+ * @param updating_resource The new resource.
+ */
+void mergeResource(Resource& old_resource, const Resource& updating_resource) {
+  if (isEmptyResource(updating_resource)) {
```
Actually, the [spec ](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.24.0/specification/resource/sdk.md#merge)somewhat mentions the case where both resources have a `schemaUrl` set but different:

>Else this is a merging error (this is the case when the Schema URL of the old and updating resources are not empty and are different). The resulting resource is undefined, and its contents are implementation-specific.

And the collector impl https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/resourcedetectionprocessor/internal/resourcedetection.go#L160 returns the old schema. Essentially I think this will result in the same behavior I have. If `updating_resource` has a schema URL but no resources, in the collector the current resource will have still the current schemaUrl and not from the updating one.

joaopgrassi replied:
> old_resource had an empty-schema, the merged resource will receive the updating_resource schema.
Yes, if `old_resource` has an empty schema and the `updating_resource` has a non-empty schema, the updating should be used. BUT in your case the `updating_resource` has also no attributes and to me, updating the old resource with the schema from the updating in this case is wrong, since the updating resource has no attributes. The schema describes the definition/conventions of the attributes it carries, and if it's empty then I'd consider it a wrong situation. I can try to bring this to the OTel spec to seek further clarification but that may take a while.

joaopgrassi replied:
@adisuissa for now, I changed the behavior in Envoy to follow the spec exactly - schema URL is merged regardless of a resource having attributes or not [512e692](https://github.com/envoyproxy/envoy/pull/29547/commits/512e692dd7eda9779b1086333e6c828432209fd2).

In the meantime I opened an issue in the OTel spec https://github.com/open-telemetry/opentelemetry-specification/issues/3710. If something there changes I will update here as a follow up.

adisuissa replied:
ACK, thanks! Please add a TODO followup in this code, so one can trace back this comment if needed in the future.

joaopgrassi replied:
Done!

joaopgrassi commented:
@adisuissa I fixed the docs CI issues. Now the new proto is present in the API docs. 

Here is how it looks:

![image](https://github.com/envoyproxy/envoy/assets/5938087/1bb710a0-974a-44b3-a056-646f409cb7c9)
![image](https://github.com/envoyproxy/envoy/assets/5938087/15c78dbe-a0be-4a9c-9215-90e9293d52a8)
![image](https://github.com/envoyproxy/envoy/assets/5938087/fcd0d1a2-5995-42c5-83f2-226ac8d7e302)
![image](https://github.com/envoyproxy/envoy/assets/5938087/f5ac3ac7-94ac-4132-b45e-5cbb78885a1f)


Sorry for the size of the PR. I wasn't sure if I should break it only adding the detectors and later using. I will keep it in mind for the next ones :)

phlax commented:
/docs

for ref @joaopgrassi you can get it to link the docs that it has built in ci - this gives a generic pr link which resolves to commit - you can link these for specific pages

joaopgrassi commented:
Hi @adisuissa sorry for the ping, but I went through the comments and it should be ready for another round :). Thanks!

ravenblackx commented:
/retest

Ping @adisuissa

adisuissa commented on test/extensions/tracers/opentelemetry/resource_detectors/resource_provider_test.cc:
```diff
@@ -0,0 +1,403 @@
+
+    EXPECT_EQ(expected_schema_url, resource.schemaUrl_);
+  }
+  // old resource schema is not empty and updating one is. should keep old schema
```
nit: please refactor these tests to individual tests instead of a single long one. This simplifies debugging when a test is broken.

joaopgrassi replied:
I wasn't sure what was the common style but I agree. Fixed in: [ca96947](https://github.com/envoyproxy/envoy/pull/29547/commits/ca96947c798cf9707d72262a091fedc078789f97)

adisuissa reviewed (commented):
Apologies for the late reply.
Here is another pass. Overall API looks good, and I think this is close to being done.
Please add an integration test that uses the driver (or a fake one).

joaopgrassi commented on source/extensions/tracers/opentelemetry/resource_detectors/resource_provider.h:
```diff
@@ -0,0 +1,44 @@
+
+#include "envoy/config/trace/v3/opentelemetry.pb.h"
+
+#include "resource_detector.h"
```
You mean I should use the full path? They are in the same folder, so I thought I didn't need to use the full location?

joaopgrassi replied:
I fixed it everywhere to use the full path. I looked at other extensions and they are doing the same.

joaopgrassi commented on source/extensions/tracers/opentelemetry/config.cc:
```diff
@@ -18,6 +18,7 @@ OpenTelemetryTracerFactory::OpenTelemetryTracerFactory()
 Tracing::DriverSharedPtr OpenTelemetryTracerFactory::createTracerDriverTyped(
     const envoy::config::trace::v3::OpenTelemetryConfig& proto_config,
     Server::Configuration::TracerFactoryContext& context) {
+
```
Done in: [c71695b](https://github.com/envoyproxy/envoy/pull/29547/commits/c71695bc595989e62523f1e8cbd192f3c493a5a1)

joaopgrassi replied:
Done in: [ca96947](https://github.com/envoyproxy/envoy/pull/29547/commits/ca96947c798cf9707d72262a091fedc078789f97)

joaopgrassi commented on api/envoy/extensions/tracers/opentelemetry/resource_detectors/v3/environment_resource_detector.proto:
```diff
@@ -0,0 +1,20 @@
+
+// Configuration for the Environment Resource detector extension.
+// The resource detector reads from the `OTEL_RESOURCE_ATTRIBUTES`
+// environment variable, as per the OpenTelemetry specification.
```
Done in c71695b

joaopgrassi commented on source/extensions/tracers/opentelemetry/opentelemetry_tracer_impl.cc:
```diff
@@ -1,3 +1,4 @@
+#include "opentelemetry_tracer_impl.h"
```
Hum that's probably a mistake. Removed in : c71695b

joaopgrassi commented on source/extensions/tracers/opentelemetry/resource_detectors/environment/BUILD:
```diff
@@ -0,0 +1,33 @@
+)
+
+envoy_cc_library(
+    name = "resource_detector_environment_lib",
```
Done in: c71695b

joaopgrassi commented on source/extensions/tracers/opentelemetry/resource_detectors/resource_detector.h:
```diff
@@ -0,0 +1,70 @@
+  virtual ~Resource() = default;
+};
+
+using ResourcePtr = std::shared_ptr<const Resource>;
```
Changed to `unique_ptr`: [ca96947](https://github.com/envoyproxy/envoy/pull/29547/commits/ca96947c798cf9707d72262a091fedc078789f97)

joaopgrassi commented on source/extensions/tracers/opentelemetry/resource_detectors/resource_detector.h:
```diff
@@ -0,0 +1,70 @@
+  virtual Resource detect() PURE;
+};
+
+using ResourceDetectorPtr = std::shared_ptr<ResourceDetector>;
```
Changed to `unique_ptr`: [ca96947](https://github.com/envoyproxy/envoy/pull/29547/commits/ca96947c798cf9707d72262a091fedc078789f97)

joaopgrassi commented on source/extensions/tracers/opentelemetry/resource_detectors/resource_provider.cc:
```diff
@@ -0,0 +1,109 @@
+namespace {
+bool isEmptyResource(const Resource& resource) { return resource.attributes_.empty(); }
+
+Resource createInitialResource(std::string service_name) {
```
Done in: [96d2bc6](https://github.com/envoyproxy/envoy/pull/29547/commits/96d2bc6683f8d850e7d211af6ec1c4b0e0fdc433)

joaopgrassi commented on source/extensions/tracers/opentelemetry/resource_detectors/environment/environment_resource_detector.cc:
```diff
@@ -0,0 +1,57 @@
+
+  std::istringstream iss(attributes_str);
+  std::string token;
+  while (std::getline(iss, token, ',')) {
```
Hum but I'm not operating on a file here, only on a string coming from the env variable. I adapted it to use `splitToken` instead. [96d2bc6](https://github.com/envoyproxy/envoy/pull/29547/commits/96d2bc6683f8d850e7d211af6ec1c4b0e0fdc433)

joaopgrassi commented on source/extensions/tracers/opentelemetry/resource_detectors/environment/environment_resource_detector.cc:
```diff
@@ -0,0 +1,57 @@
+  while (std::getline(iss, token, ',')) {
+    size_t pos = token.find('=');
+    std::string key = token.substr(0, pos);
+    std::string value = token.substr(pos + 1);
```
Done in: [96d2bc6](https://github.com/envoyproxy/envoy/pull/29547/commits/96d2bc6683f8d850e7d211af6ec1c4b0e0fdc433) (now also using `splitToken`)

joaopgrassi commented on source/extensions/tracers/opentelemetry/resource_detectors/environment/environment_resource_detector.cc:
```diff
@@ -0,0 +1,57 @@
+  std::istringstream iss(attributes_str);
+  std::string token;
+  while (std::getline(iss, token, ',')) {
+    size_t pos = token.find('=');
```
Done in: [96d2bc6](https://github.com/envoyproxy/envoy/pull/29547/commits/96d2bc6683f8d850e7d211af6ec1c4b0e0fdc433) (add a check, bit different now because of I'm using `splitToken`)

joaopgrassi commented:
@adisuissa  thanks for the review! I think I addressed everything. 

About this: 

>Please add an integration test that uses the driver (or a fake one).

Sorry I'm not sure I got it. I modified the existing driver tests (`opentelemetry_tracer_impl_test.cc`) to include a "mock" detector and assert the exported OTLP message contains the resource attributes https://github.com/envoyproxy/envoy/pull/29547/files#diff-75d4e9c804c8a50a7b7f020fa7b5de184462733e2fad07e86c415e4b0ce879bd 

Is that maybe what you meant or something else?

adisuissa reviewed (commented):
Thanks for working on this.
OOC is there a plan to add an integration test (that tests a running Envoy with an emulated OTel tracer server)?

joaopgrassi commented on source/extensions/tracers/opentelemetry/resource_detectors/environment/config.cc:
```diff
@@ -0,0 +1,35 @@
+ResourceDetectorPtr EnvironmentResourceDetectorFactory::createResourceDetector(
+    const Protobuf::Message& message, Server::Configuration::TracerFactoryContext& context) {
+
+  auto mptr = Envoy::Config::Utility::translateAnyToFactoryConfig(
```
Hum I based my self on other extensions, such as the `original_ip_detection`. I tried removing it and the tests fail now with `C++ exception with description "std::bad_cast" thrown in the test body.`

joaopgrassi replied:
That is already the invocation _on_ the factory object. The factory creation is using `getFactory` as you pointed here: https://github.com/envoyproxy/envoy/pull/29547/files#diff-74b41d6afb01dd775d2adcf9b25504fdd611ab3cb9d35b1c339fec8a5298ac2bR87.

The factory is then used to create "any detector" based on the config type, `createResourceDetector` https://github.com/envoyproxy/envoy/pull/29547/files#diff-74b41d6afb01dd775d2adcf9b25504fdd611ab3cb9d35b1c339fec8a5298ac2bR94. 

Then it ends up in here `environment/config.cc` where we create the actual type `EnvironmentResourceDetector`. And for that we need the proper casted config type

joaopgrassi commented on source/extensions/tracers/opentelemetry/opentelemetry_tracer_impl.cc:
```diff
@@ -22,12 +24,22 @@ namespace OpenTelemetry {
   auto& factory_context = context.serverFactoryContext();
+
+  Resource resource = resource_provider.getResource(opentelemetry_config, context);
+  ResourcePtr resource_ptr = std::make_shared<Resource>(std::move(resource));
```
Actually I think it should be shared. I changed it here https://github.com/envoyproxy/envoy/pull/29547#discussion_r1337427289 but I think it was a mistake. `resource_ptr` is captured by the lambda below so I think it's better to use a shared one. I changed it back again and used your suggestions for the name `ResourceConstSharedPtr` [d79e015](https://github.com/envoyproxy/envoy/pull/29547/commits/d79e015238fb8b1b115fcf24aafeeab897cf5801)

adisuissa reviewed (commented):
Apologies for the late response.
I've left a minor comment, and a followup to the discussion on rejecting/ignoring an invalid config.

jmarantz commented:
/wait

htuch commented:
> I'm inclined towards rejecting a wrong config, to keep it with the rest of Envoy configs (and also because if in the future one decides to changing this to ignoring the config, it will break less users).
@htuch thoughts on this?

Makes sense to me.

joaopgrassi commented:
> > I'm inclined towards rejecting a wrong config, to keep it with the rest of Envoy configs (and also because if in the future one decides to changing this to ignoring the config, it will break less users).
> > @htuch thoughts on this?
> 
> Makes sense to me.

@adisuissa @htuch I addressed this now. Envoy will not start if the resource is not "detected" from the env variable. Let me know what you think!

htuch commented:
That aspect looks fine to me now.

joaopgrassi commented:
@adisuissa hi! I went through the comments, it should be good for another look. Thank you!

adisuissa reviewed (approved):
LGTM, thanks!
/assign-from @envoyproxy/senior-maintainers

RyanTheOptimist commented on api/envoy/extensions/tracers/opentelemetry/resource_detectors/v3/environment_resource_detector.proto:
```diff
@@ -0,0 +1,25 @@
+// [#protodoc-title: Environment Resource Detector config]
+
+// Configuration for the Environment Resource detector extension.
+// The resource detector reads from the ``OTEL_RESOURCE_ATTRIBUTES``
```
Did you consider having this information in the config instead of in an environment variable?

RyanTheOptimist commented on source/extensions/tracers/opentelemetry/resource_detectors/resource_detector.h:
```diff
@@ -0,0 +1,70 @@
+
+using ResourceAttributes = std::map<std::string, std::string>;
+
+struct Resource {
```
nit: please add a short comment to describe what this represents.

joaopgrassi replied:
Done in [a374106](https://github.com/envoyproxy/envoy/pull/29547/commits/a374106c23c6f855cd0bdba275707c3449d20118)

RyanTheOptimist reviewed (approved):
Looks great. Two small comments.

RyanTheOptimist commented:
Looks like a formatting issue?

./source/extensions/tracers/opentelemetry/resource_detectors/resource_detector.h:16: * @brief A strig key-value map that stores the resource attributes.
                                                                                                ^
Checked 4800 file(s) and 132399 comment(s), found 1 error(s).
ERROR: spell check failed. Run 'tools/spelling/check_spelling_pedantic.py fix and/or add new words to tools/spelling/spelling_dictionary.txt'

joaopgrassi commented:
@RyanTheOptimist sorry, it was a typo. fixed!

RyanTheOptimist commented:
Looks like this needs a main merge
/wait

joaopgrassi commented:
@RyanTheOptimist update and solved the conflict. I gave write permission to maintainers on the branch/PR, in case it happens again. Should be good to go!

joaopgrassi commented:
/retest
